packages/hardhat-core/src/internal/hardhat-network/provider/fork/ForkStateManager.ts

```diff
@@ -50,6 +51,15 @@ export class ForkStateManager implements PStateManager {
     this._state = ImmutableMap();
   }
 
+  /**
+   * The forked block needs special handling because the genesis accounts are added to it.
+   * This is meant to be called when the initial state is final, so that _initialStateRoot
+   * points to this starting fork state.
+   */
+  public updateInitialStateRoot() {
+    this._stateRootToState.set(this._initialStateRoot, this._state);
+  }
```

Comment:
Instead of adding this function and a long comment explaining its purpose I would put all of the initialization logic inside the FSM constructor. In other words, move this:
https://github.com/nomiclabs/hardhat/blob/b21474ce1df891edd7fb7b94f8bbc079c652b85a/packages/hardhat-core/src/internal/hardhat-network/provider/utils/putGenesisAccounts.ts#L10-L13
and make `genesisAccounts` a constructor parameter.

Reply:
Btw, `putAccount` is async only for the purpose of compliance with the `PStateManager` interface. I'd add a private synchronous `_putAccount` method and call it in the constructor and in the original public `putAccount`.